### PR TITLE
Fixes unwanted floating chat duplication bug

### LINKED ISSUE #5062: Chutes Bugged
Submitter: MorganTerim
**Describe the bug**
After getting out of a disposals chute which has an item in it while you are talking, your text appears on both your head and the chute at the same time

**To Reproduce**
1. Put an item in a disposals chute
2. Get in the disposals chute
3. Say something and while the text is coming out from your head ,leave the chute
4. Have your text both on your head and the chute

**Expected behavior**
There shouldnt be text on my head and chute at the same time

**Screenshots**
Not exactly sure how i can add it here since i dont use any screenshot to link software if you contact me on discord Bepis#0121 i will be able to provide them

**Additional context**
None

Reported on: Goonstation 1 Classic: Heisenbee 2021-06-08 07:50:16

### LINKED ISSUE #5068: clothing booth copies overhead text
[BUG]

**Describe the bug**
occasionally, using the clothing booth will cause your overhead text to replicate as if you were still inside the booth. it still works for your own overhead, so you just have two talking at once

**To Reproduce**
1. step into a clothing booth
2. step out of the clothing booth
3. close the ui and reopen it by clicking the booth again (this may not effect it, its just something i do often when the bug appears)
3. observe

ive also seen this happen with pods, but i am unsure how to go about replicating it, and it seems much more common with the clothing booth

### LINKED ISSUE #5072: Floating chat bug when exiting a pod
Submitter: metricduck
**Describe the bug**
After exiting a pod for a while after either talking or emoting (including automatic emotes including shiver) will appear over both you and the pod you just got out of.

**To Reproduce**
1. Get out of pod.
2. Talk.
3. Text appears over both the spaceman and the pod.

**Expected behavior**
Floaty text appears only over the spaceman.

**Screenshots**
-

**Additional context**
I've only noticed the bug in the last 48 hours or so.

Reported on: Goonstation 1 Classic: Heisenbee 2021-06-08 17:29:51

### LINKED ISSUE #5102: Visual Bug, Duplicaiton of voices
Submitter: orgslug
As clear as I can tell its a visual error where my voice was appearing in multiple locations (in this case extra pods, both regular and miniputt).

To Reproduce, I'm not sure.  There were two of us in the pod and several other pods warped in on top of us. 

**Expected behavior**
Nothing, visual bug.

Not usually someone who reports bugs so my description is lacking.  It was just a strange thing that I'm sure you've seen before. 

Reported on: Goonstation 3 Roleplay: Morty 2021-06-10 11:37:36

### LINKED ISSUE #5108: Double Talk
Submitter: k3ynan
teleport out of a jail artifact with a hand held teleport artifact

**To Reproduce**
Aquire artifact powers of bulky jail and handheld teleporter
Jail yourself and teleport out.
You talk in both places
Uh oh

**Expected behavior**
Teleport out nothing else happens

**Screenshots**

**Additional context**

Reported on: Goonstation 3 Roleplay: Morty 2021-06-10 21:05:25

### PR DESCRIPTION
**Pull request by eX-n0x**
[BUG] [EXPERIMENTAL] [HELP WANTED]
## About the PR 
This essentially instantly resets an object that the player gets out of so that their associated floating text doesn't stay stuck to said object. This fixes the issue of their floating text being duplicated as something gets "caught" onto certain objects that only resets when that player walks far enough away from that object. This just takes care of that "walking away" the second they get out.

Experimental/Help Wanted tags are there because: 
- ~~I am unsure how it affects other clients such as if multiple people are in a pod together and one person hops out causing the instant reset.~~ After some further testing I feel that it shouldn't negatively affect other players just due to how the invisibility var works and how quickly it is changed. Objs still worked like normal even when invisible and doesn't affect the player's ability to perform the actions they want.

- Also not sure about what exactly is getting caught on certain objects and if there is a better way to get rid of that "leftover" however this fix does work in resolving the issue.

## Why's this needed? 
Fixes #5108, fixes #5068, fixes #5072, fixes #5102, fixes #5062, and whatever other ones are out there.
None of these instances should have the same floating text in multiple places that the player doesn't currently occupy.

## Changelog

```changelog
(u)eX.n0x
(+)Floating chat no longer gets duplicated when exiting certain objects.
```

### CODE REVIEW

**Comment by zjdtmkhzt** [comment]
I see you only reset on /obj/machinery. This also happens on artifact prisons though (and presumably other things that can have people inside them)?
So, would this fix the bug for those?

**Comment by eX-n0x** [comment]
> I see you only reset on /obj/machinery. This also happens on artifact prisons though (and presumably other things that can have people inside them)?
> So, would this fix the bug for those?
> 
> EDIT: Also, no idea what's up with Destiny failing. Maybe it was just random and another commit might fix it on the next run.

Should work on anything that allows for an invisibility change. Was just trying to limit it as much as possible to avoid any complications with calling it on anything since, at least in my testing, not everything gives you this bug when you enter and exit them. Not only that but this is more of a bandaid fix since I can't find what is actually happening even after hours of debugging and checking crap like vis stuff and pools.

Also, rip me forgot about artifacts, I'll get that set up in a little...

File: code/mob/living.dm [hunk 1]
```diff
@@ -1344,6 +1344,11 @@ var/global/icon/human_static_base_idiocy_bullshit_crap = icon('icons/mob/human.d
 		if(src.chat_text.vis_locs.len)
 			var/atom/movable/AM = src.chat_text.vis_locs[1]
 			AM.vis_contents -= src.chat_text
+
+			//Fixes spooky floating text dup with spooky code
+			if((istype(AM, /obj/machinery) || istype(AM, /obj/artifact)) && src.client)
+				AM.invisibility = 101
+				AM.invisibility = 0
```
Comment (ZeWaka) [inline code/mob/living.dm:1351]: i would be very surprised if this is different from just setting to 0.

Comment (eX-n0x) [reply code/mob/living.dm:1351]: Doesn't seem to work if it isn't made fully invisible first, (just tested). The duplication issue goes away when the player gets far enough from the object where they can't see it, but if you make it fully invisible it also resolves that issue.

**Comment by Sovexe** [comment]
This can happen on basically any atom and is easily reproduced using noclip

Not a fan of this bandaid fix since it does nothing to actually remedy the underlying issue

Pretty certain it didnt used to happen. Seems to the it's either a byond bug or something that we did. I'd say we'd be better off trying lower versions and old commits to find the source.

**Comment by eX-n0x** [comment]
> This can happen on basically any atom and is easily reproduced using noclip
> 
> Not a fan of this bandaid fix since it does nothing to actually remedy the underlying issue. Additionally setting invis to 0 is not always correct, so this will cause bugs.
> 
> Pretty certain it didnt used to happen. Seems to me it's either a byond bug or something that we did. I'd say we'd be better off trying lower versions and old commits to find the source.

Thanks for the reminder that I need to reset the obj back to whatever its original inv was. Mainly this is just targeting what players were reporting most that were giving them errors, mainly machines and artifacts. I'll try testing a clean master with other builds of BYOND since I did spend a lot of time already trying to trace through the call stack to see if anything popped out at me about being an error but...everything seemed correct.

By the way, if this is just a BYOND error then would this PR even be used, or would it be more of a just "wait and see" if it gets fixed later on since I wouldn't think that it would be better to downgrade the BYOND build for the future?

**Comment by Sovexe** [comment]
If it's a byond error we need to figure out how code wise what to do to cause it so we can submit a test case and beta bug report to lummox

File: code/mob/living.dm [hunk 2]
```diff
@@ -1344,6 +1344,12 @@ var/global/icon/human_static_base_idiocy_bullshit_crap = icon('icons/mob/human.d
 		if(src.chat_text.vis_locs.len)
 			var/atom/movable/AM = src.chat_text.vis_locs[1]
 			AM.vis_contents -= src.chat_text
+
+			//Fixes spooky floating text dup with spooky code
+			if((istype(AM, /obj/machinery) || istype(AM, /obj/artifact)) && src.client)
+				var/atom/movable/invtemp = AM.invisibility
```
Comment (ZeWaka) [inline code/mob/living.dm:1350]:
```suggestion
				var/invtemp = AM.invisibility
```
definitely not

**Review by ZeWaka — CHANGES_REQUESTED** [review]

**Comment by eX-n0x** [comment]
> If it's a byond error we need to figure out code wise what to do to cause it so that we can submit a test case and beta bug report to lummox

Just tested building the master in a VM. Seems the issue was introduced in build _.1555 and is still a regression in .1556_. **.1554 does not have this issue** and I made sure to double-check both builds. I only ever started messing with BYOND coding just a few days ago so I'm not sure how to even create a proper test case besides just copying the bare essentials from Gooncode and showing that off.

I really do feel that it's some issue with either vis_locs or vis_contents not having their garbage collected when the targets switch from whatever the player is inside of back to a turf since debugging shows that they get properly switched to the right atom when the set_loc func is called. It just seems that removing an atom from one of the two vis lists isn't enough anymore to get the game to recognize that chat_maptext shouldn't come from the previous atom. Turning that object completely invisible just seems to get the game to finally clear out whatever it's still holding onto and to properly dispose of that old garbage data.

**Comment by Sovexe** [comment]
http://www.byond.com/forum/post/2692609

**Comment by eX-n0x** [comment]
> http://www.byond.com/forum/post/2692609

Thank you for submitting that, hope I was at least somewhat helpful. Also apologies for not submitting that report myself, just wasn't sure if I was right or not from guessing nor what to do for a nice small test case.

**Comment by eX-n0x** [comment]
Did a quick test and it seems the latest BYOND update fixes these issues based on the latest master. Can probably close those bug reports.

**Pull request closed without merging**